File: aihuoke/ai-hunter-root/backend/auth_sso.py
"""Sub2API SSO and three-day HttpOnly sessions for AI获客."""

from __future__ import annotations

import base64
import binascii
import hashlib
import hmac
import json
import math
import os
import secrets
import sqlite3
import time
from contextvars import ContextVar
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import httpx
from fastapi import APIRouter, HTTPException, Request, Response
from fastapi.responses import RedirectResponse
from urllib.parse import urlsplit, urlunsplit

SESSION_COOKIE = "aihuoke_session"
SESSION_TTL_SECONDS = 3 * 24 * 60 * 60
_CURRENT_SUBJECT: ContextVar[str] = ContextVar("aihuoke_sub2api_subject", default="")
router = APIRouter(prefix="/api/auth", tags=["auth"])
balance_router = APIRouter(prefix="/api/sub2api", tags=["sub2api"])


def _load_local_env() -> None:
    """Load ``backend/.env`` for non-Docker launches without overriding env.

    Docker/systemd deployments inject their environment directly.  Reading
    the local file here keeps the SSO and relay variables consistent with the
    settings layer when the README's ``cp .env.example .env`` workflow is
    used, while preserving explicitly injected deployment values.
    """
    env_path = Path(__file__).resolve().parent / ".env"
    try:
        lines = env_path.read_text(encoding="utf-8").splitlines()
    except OSError:
        return
    for line in lines:
        text = line.strip()
        if not text or text.startswith("#") or "=" not in text:
            continue
        key, value = text.split("=", 1)
        key = key.strip()
        value = value.strip().strip('"').strip("'")
        if key and key not in os.environ:
            os.environ[key] = value


_load_local_env()


def _has_configured_secret(value: object) -> bool:
    normalized = str(value or "").strip().lower()
    if not normalized:
        return False
    return not (
        normalized in {"changeme", "change-me", "replace-me", "your-key", "your_api_key", "your-api-key", "xxx", "sk-xxx"}
        or normalized.startswith(("your_", "your-", "replace-with-", "example-", "sk-super-"))
    )


def has_configured_secret(value: object) -> bool:
    """Return whether a secret is real rather than an example placeholder.

    Keep the placeholder policy in one place so status endpoints do not claim
    that the checked-in ``.env.example`` values have enabled the relay.
    """
    return _has_configured_secret(value)


@dataclass(frozen=True)
class Identity:
    id: int
    subject: str
    username: str = ""
    display_name: str = ""


def _db_path() -> Path:
    raw = os.getenv("AIHUOKE_AUTH_DB", "").strip()
    path = Path(raw) if raw else Path(__file__).resolve().parent / "data" / "auth.db"
    path.parent.mkdir(parents=True, exist_ok=True)
    return path


def _connect() -> sqlite3.Connection:
    conn = sqlite3.connect(_db_path(), timeout=5)
    conn.row_factory = sqlite3.Row
    conn.executescript(
        """
        CREATE TABLE IF NOT EXISTS identities (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          issuer TEXT NOT NULL,
          subject TEXT NOT NULL,
          username TEXT NOT NULL DEFAULT '',
          display_name TEXT NOT NULL DEFAULT '',
          created_at INTEGER NOT NULL,
          updated_at INTEGER NOT NULL,
          UNIQUE (issuer, subject)
        );
        CREATE TABLE IF NOT EXISTS consumed_tickets (
          jti TEXT PRIMARY KEY,
          expires_at INTEGER NOT NULL,
          consumed_at INTEGER NOT NULL
        );
        CREATE TABLE IF NOT EXISTS sessions (
          token_hash TEXT PRIMARY KEY,
          identity_id INTEGER NOT NULL REFERENCES identities(id),
          expires_at INTEGER NOT NULL,
          created_at INTEGER NOT NULL
        );
        """
    )
    conn.commit()
    return conn


def _decode(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def verify_ticket(raw: str, now: int | None = None) -> dict[str, Any]:
    secret = os.getenv("SUB2API_SSO_SECRET", "").strip()
    if not _has_configured_secret(secret) or len(secret) < 32:
        raise ValueError("SSO is not configured")
    if not isinstance(raw, str) or not raw or len(raw) > 8192:
        raise ValueError("invalid ticket")
    try:
        encoded, signature = raw.split(".", 1)
        expected = hmac.new(secret.encode(), encoded.encode(), hashlib.sha256).digest()
        if not hmac.compare_digest(expected, _decode(signature)):
            raise ValueError("invalid signature")
        payload = json.loads(_decode(encoded))
    except (ValueError, TypeError, UnicodeDecodeError, json.JSONDecodeError, binascii.Error) as exc:
        raise ValueError("invalid ticket") from exc
    if not isinstance(payload, dict) or payload.get("iss") != "sub2api" or payload.get("aud") != "aihuoke":
        raise ValueError("invalid audience")
    if any(key in payload for key in ("rk", "api_key", "super_key")):
        raise ValueError("credential material is not allowed")
    subject, jti = payload.get("sub"), payload.get("jti")
    if not isinstance(subject, str) or not subject.strip() or len(subject) > 160 or not isinstance(jti, str) or not jti.strip() or len(jti) > 200:
        raise ValueError("invalid identity")
    issued, expires = payload.get("iat"), payload.get("exp")
    current = int(time.time()) if now is None else int(now)
    if type(issued) is not int or type(expires) is not int or issued <= 0 or expires <= issued or expires <= current or expires - issued > 120 or issued > current + 30:
        raise ValueError("expired ticket")
    return payload


def safe_next(value: object) -> str:
    if not isinstance(value, str) or not value.startswith("/") or value.startswith("//") or "\\" in value or len(value) > 2048 or any(ord(c) < 32 for c in value):
        return "/"
    return value


def consume_ticket(payload: dict[str, Any]) -> Identity:
    now = int(time.time())
    conn = _connect()
    try:
        conn.execute("DELETE FROM consumed_tickets WHERE expires_at < ?", (now,))
        try:
            conn.execute("INSERT INTO consumed_tickets(jti, expires_at, consumed_at) VALUES (?, ?, ?)", (payload["jti"], int(payload["exp"]), now))
        except sqlite3.IntegrityError as exc:
            conn.rollback()
            raise ValueError("ticket already used") from exc
        row = conn.execute("SELECT * FROM identities WHERE issuer = ? AND subject = ?", ("sub2api", payload["sub"])).fetchone()
        if row is None:
            conn.execute("INSERT INTO identities(issuer, subject, username, display_name, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)", ("sub2api", payload["sub"], str(payload.get("username") or ""), str(payload.get("displayName") or ""), now, now))
            identity_id = int(conn.execute("SELECT last_insert_rowid()").fetchone()[0])
        else:
            identity_id = int(row["id"])
            conn.execute("UPDATE identities SET username = ?, display_name = ?, updated_at = ? WHERE id = ?", (str(payload.get("username") or row["username"]), str(payload.get("displayName") or row["display_name"]), now, identity_id))
        conn.commit()
        row = conn.execute("SELECT * FROM identities WHERE id = ?", (identity_id,)).fetchone()
        assert row is not None
        return Identity(int(row["id"]), str(row["subject"]), str(row["username"]), str(row["display_name"]))
    finally:
        conn.close()


def create_session(identity: Identity, response: Response, request: Request) -> None:
    token = secrets.token_urlsafe(32)
    now = int(time.time())
    conn = _connect()
    try:
        conn.execute("INSERT INTO sessions(token_hash, identity_id, expires_at, created_at) VALUES (?, ?, ?, ?)", (_hash(token), identity.id, now + SESSION_TTL_SECONDS, now))
        conn.commit()
    finally:
        conn.close()
    forwarded = request.headers.get("x-forwarded-proto", request.url.scheme).split(",", 1)[0].strip().lower()
    response.set_cookie(SESSION_COOKIE, token, max_age=SESSION_TTL_SECONDS, httponly=True, secure=forwarded == "https", samesite="lax", path="/")


def get_identity(request: Request) -> Identity | None:
    token = request.cookies.get(SESSION_COOKIE, "").strip()
    if not token:
        return None
    conn = _connect()
    try:
        row = conn.execute("SELECT i.* FROM sessions s JOIN identities i ON i.id = s.identity_id WHERE s.token_hash = ? AND s.expires_at >= ?", (_hash(token), int(time.time()))).fetchone()
        return None if row is None else Identity(int(row["id"]), str(row["subject"]), str(row["username"]), str(row["display_name"]))
    finally:
        conn.close()


def require_identity(request: Request) -> Identity:
    identity = get_identity(request)
    if identity is None:
        raise HTTPException(status_code=401, detail="Authentication required")
    return identity


def set_current_subject(subject: str):
    return _CURRENT_SUBJECT.set(str(subject or "").strip())


def reset_current_subject(token: object) -> None:
    _CURRENT_SUBJECT.reset(token)  # type: ignore[arg-type]


def current_subject() -> str:
    return _CURRENT_SUBJECT.get()


def managed() -> bool:
    # Fail closed when a deployment explicitly supplied either integration
    # variable, even if it still contains the readable value from an example
    # file.  Treating that value as "local mode" would silently disable SSO
    # and expose the standalone, unauthenticated surface.
    return bool(
        str(os.getenv("SUB2API_SSO_SECRET", "") or "").strip()
        or str(os.getenv("SUB2API_APP_CREDENTIAL", "") or "").strip()
    )


def scoped_email_db_path(base_path: str, subject: str | None = None) -> str:
    """Return the email database owned by one Sub2API subject.

    Local mode deliberately keeps the historical single database path.  In a
    managed deployment every authenticated subject gets a separate SQLite
    file, so account credentials, suppression lists, campaigns and worker
    state cannot cross an identity boundary.  The subject is represented only
    by a one-way directory name and is never written to a browser-visible
    value or a filename.
    """
    raw_path = str(base_path or "email_automation.db").strip() or "email_automation.db"
    owner = str(subject if subject is not None else current_subject()).strip()
    if not managed() or not owner:
        return raw_path
    path = Path(raw_path)
    digest = hashlib.sha256(owner.encode("utf-8")).hexdigest()[:32]
    directory = path.parent / f"{path.stem}.users"
    return str(directory / f"{digest}{path.suffix or '.db'}")


def default_email_account_id(subject: str | None = None) -> str:
    """Return a collision-free default account id for a managed subject."""
    owner = str(subject if subject is not None else current_subject()).strip()
    if not managed() or not owner:
        return "default"
    return f"default-{hashlib.sha256(owner.encode('utf-8')).hexdigest()[:24]}"


def known_subjects() -> set[str]:
    """Read known SSO subjects for maintenance/public unsubscribe workers."""
    if not managed():
        return set()
    try:
        conn = _connect()
        try:
            rows = conn.execute("SELECT subject FROM identities WHERE issuer = 'sub2api'").fetchall()
        finally:
            conn.close()
        return {str(row["subject"] or "").strip() for row in rows if str(row["subject"] or "").strip()}
    except sqlite3.Error:
        return set()


def relay_base_url() -> str:
    raw = (os.getenv("SUB2API_RELAY_BASE_URL") or os.getenv("LINK") or "").strip()
    if raw and "://" not in raw:
        raw = "http://" + raw
    raw = raw.rstrip("/")
    return raw if raw.endswith("/v1") else (raw + "/v1" if raw else "")


def satellite_headers(*, required: bool = True) -> dict[str, str]:
    subject = current_subject()
    credential = os.getenv("SUB2API_APP_CREDENTIAL", "").strip()
    if required and (not subject or not _has_configured_secret(credential)):
        raise RuntimeError("Sub2API satellite credential and session subject are required")
    if not subject or not _has_configured_secret(credential):
        return {}
    return {"Authorization": f"Bearer {credential}", "X-Sub2API-On-Behalf-Of": subject, "X-Sub2API-Satellite": "aihuoke"}


def sub2api_purchase_url() -> str | None:
    raw = os.getenv("LINK", "").strip()
    if not raw:
        return None
    if "://" not in raw:
        raw = "http://" + raw
    try:
        parts = urlsplit(raw)
        if parts.scheme not in {"http", "https"} or not parts.hostname or parts.username or parts.password or parts.query or parts.fragment:
            return None
        return urlunsplit((parts.scheme, parts.netloc, "/purchase", "", ""))
    except ValueError:
        return None


@balance_router.get("/balance")
async def sub2api_balance(response: Response):
    response.headers["Cache-Control"] = "no-store"
    subject = current_subject().strip()
    if not managed() or not subject:
        raise HTTPException(status_code=401, detail="Authentication required")
    base = relay_base_url()
    if not base:
        raise HTTPException(status_code=503, detail="Sub2API is not configured")
    try:
        headers = satellite_headers()
    except RuntimeError as exc:
        raise HTTPException(status_code=503, detail="Sub2API satellite is not configured") from exc
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            upstream = await client.get(f"{base}/sub2api/balance", headers=headers)
        if upstream.status_code >= 400:
            status = upstream.status_code if upstream.status_code in {401, 503} else 502
            raise HTTPException(status_code=status, detail="Balance is unavailable")
        data = upstream.json()
        balance = float(data["balance"])
        if not math.isfinite(balance):
            raise ValueError("invalid balance")
    except HTTPException:
        raise
    except (httpx.HTTPError, ValueError, KeyError, TypeError) as exc:
        raise HTTPException(status_code=502, detail="Balance is unavailable") from exc
    return {"balance": balance, "recharge_url": sub2api_purchase_url()}


@router.get("/sso/callback")
async def sso_callback(request: Request):
    try:
        payload = verify_ticket(request.query_params.get("ticket", ""))
        identity = consume_ticket(payload)
    except (ValueError, TypeError) as exc:
        if str(exc) == "SSO is not configured":
            raise HTTPException(status_code=503, detail="SSO is not configured") from exc
        return RedirectResponse("/?error=sso_failed", status_code=303, headers={"Cache-Control": "no-store", "Referrer-Policy": "no-referrer"})
    response = RedirectResponse(safe_next(payload.get("next")), status_code=303, headers={"Cache-Control": "no-store", "Referrer-Policy": "no-referrer"})
    create_session(identity, response, request)
    return response


@router.get("/me")
async def me(request: Request):
    identity = require_identity(request)
    return {"id": identity.id, "subject": identity.subject, "username": identity.username, "display_name": identity.display_name}


def _hash(value: str) -> str:
    return hashlib.sha256(value.encode()).hexdigest()

File: ai3d/server/auth-sso.mjs
import { createHmac, timingSafeEqual, randomBytes, createHash } from 'node:crypto'
import { readFile, writeFile, mkdir, rename } from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { hasConfiguredSecret } from './config.mjs'

export const SESSION_COOKIE = 'ai3d_session'
export const SESSION_TTL_SECONDS = 3 * 24 * 60 * 60

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..')
const DB_FILE = path.resolve(ROOT, process.env.AI3D_AUTH_DB || '.auth/sessions.json')
let taskOwnerWrite = Promise.resolve()
let dbMutation = Promise.resolve()

function decodePart(value) {
  const normalized = String(value || '').replace(/-/g, '+').replace(/_/g, '/') + '='.repeat((4 - String(value || '').length % 4) % 4)
  return Buffer.from(normalized, 'base64')
}

function encodePart(value) {
  return Buffer.from(value).toString('base64url')
}

function hash(value) {
  return createHash('sha256').update(String(value)).digest('hex')
}

async function readDb() {
  try {
    const parsed = JSON.parse(await readFile(DB_FILE, 'utf8'))
    return {
      identities: parsed.identities && typeof parsed.identities === 'object' ? parsed.identities : {},
      consumed: parsed.consumed && typeof parsed.consumed === 'object' ? parsed.consumed : {},
      sessions: parsed.sessions && typeof parsed.sessions === 'object' ? parsed.sessions : {},
      taskOwners: parsed.taskOwners && typeof parsed.taskOwners === 'object' ? parsed.taskOwners : {},
    }
  } catch {
    return { identities: {}, consumed: {}, sessions: {}, taskOwners: {} }
  }
}

async function writeDb(db) {
  await mkdir(path.dirname(DB_FILE), { recursive: true })
  const temp = `${DB_FILE}.${process.pid}.${Date.now()}.tmp`
  await writeFile(temp, JSON.stringify(db), 'utf8')
  await rename(temp, DB_FILE)
}

// All auth/task-owner updates are read-modify-write operations on one JSON
// file. Serializing the whole transaction prevents concurrent SSO callbacks,
// session creation, and task-owner writes from silently overwriting each
// other's identities or ownership records.
function withDbMutation(mutator) {
  const operation = dbMutation.then(async () => {
    const db = await readDb()
    const result = await mutator(db)
    await writeDb(db)
    return result
  })
  dbMutation = operation.catch(() => {})
  return operation
}

export function safeNext(value) {
  const next = typeof value === 'string' ? value.trim() : ''
  const hasControlCharacter = [...next].some((character) => character.charCodeAt(0) < 0x20)
  if (!next || !next.startsWith('/') || next.startsWith('//') || next.includes('\\') || next.length > 2048 || hasControlCharacter) return '/'
  return next
}

export function verifyTicket(raw, now = Math.floor(Date.now() / 1000)) {
  const secret = String(process.env.SUB2API_SSO_SECRET || '').trim()
  if (!hasConfiguredSecret(secret) || secret.length < 32) throw Object.assign(new Error('SSO is not configured'), { status: 503 })
  if (typeof raw !== 'string' || raw.length > 8192) throw Object.assign(new Error('Invalid SSO ticket'), { status: 400 })
  let encoded, signature, payload
  try {
    const parts = raw.split('.')
    if (parts.length !== 2) throw new Error('malformed')
    encoded = parts[0]
    signature = parts[1]
    if (!encoded || !signature) throw new Error('malformed')
    const expected = createHmac('sha256', secret).update(encoded).digest()
    const actual = decodePart(signature)
    if (actual.length !== expected.length || !timingSafeEqual(actual, expected)) throw new Error('signature')
    payload = JSON.parse(decodePart(encoded).toString('utf8'))
  } catch (error) {
    throw Object.assign(new Error('Invalid SSO ticket'), { status: 400, cause: error })
  }
  if (!payload || typeof payload !== 'object' || payload.iss !== 'sub2api' || payload.aud !== 'ai3d') throw Object.assign(new Error('Invalid SSO audience'), { status: 400 })
  if ('rk' in payload || 'api_key' in payload || 'super_key' in payload) throw Object.assign(new Error('Credential material is not allowed'), { status: 400 })
  if (typeof payload.sub !== 'string' || !payload.sub.trim() || payload.sub.length > 160 || typeof payload.jti !== 'string' || !payload.jti.trim()) throw Object.assign(new Error('Invalid SSO identity'), { status: 400 })
  if (!Number.isInteger(payload.iat) || !Number.isInteger(payload.exp) || payload.iat <= 0 || payload.exp <= payload.iat || payload.exp <= now || payload.exp - payload.iat > 120 || payload.iat > now + 30) throw Object.assign(new Error('Expired SSO ticket'), { status: 400 })
  return payload
}

export async function consumeTicket(payload) {
  return withDbMutation(async (db) => {
    const now = Math.floor(Date.now() / 1000)
    for (const [key, expiry] of Object.entries(db.consumed)) if (Number(expiry) < now) delete db.consumed[key]
    if (db.consumed[payload.jti]) throw Object.assign(new Error('SSO ticket already used'), { status: 400 })
    db.consumed[payload.jti] = payload.exp
    const identityKey = `sub2api:${payload.sub}`
    const identity = db.identities[identityKey] || {
      id: hash(identityKey).slice(0, 32),
      issuer: 'sub2api',
      subject: payload.sub,
      username: String(payload.username || ''),
      displayName: String(payload.displayName || ''),
      createdAt: now,
    }
    identity.username = String(payload.username || identity.username || '')
    identity.displayName = String(payload.displayName || identity.displayName || '')
    identity.updatedAt = now
    db.identities[identityKey] = identity
    return identity
  })
}

export async function createSession(identity, request, response) {
  const token = randomBytes(32).toString('base64url')
  const now = Math.floor(Date.now() / 1000)
  await withDbMutation(async (db) => {
    db.sessions[hash(token)] = { identityKey: `sub2api:${identity.subject}`, expiresAt: now + SESSION_TTL_SECONDS, createdAt: now }
  })
  const forwarded = String(request.headers['x-forwarded-proto'] || '').split(',')[0].trim().toLowerCase()
  const secure = request.socket?.encrypted || forwarded === 'https'
  const attrs = [`${SESSION_COOKIE}=${encodeURIComponent(token)}`, 'Path=/', 'HttpOnly', 'SameSite=Lax', `Max-Age=${SESSION_TTL_SECONDS}`]
  if (secure) attrs.push('Secure')
  response.setHeader('Set-Cookie', attrs.join('; '))
}

function readCookie(request, name) {
  const raw = String(request.headers.cookie || '')
  const match = raw.match(new RegExp(`(?:^|;\\s*)${name}=([^;]*)`))
  if (!match) return ''
  try {
    return decodeURIComponent(match[1])
  } catch {
    return ''
  }
}

export async function getIdentity(request) {
  const token = readCookie(request, SESSION_COOKIE)
  if (!token) return null
  const db = await readDb()
  const session = db.sessions[hash(token)]
  if (!session || Number(session.expiresAt) <= Math.floor(Date.now() / 1000)) return null
  return db.identities[session.identityKey] || null
}

export async function requireIdentity(request) {
  const identity = await getIdentity(request)
  if (!identity) throw Object.assign(new Error('Authentication required'), { status: 401 })
  return identity
}

// Generation task ownership is persisted alongside the local auth database so
// a process restart cannot turn an authenticated user's task id into a public
// lookup. Values are kept deliberately minimal: only the Sub2API subject is
// stored, never a credential or request payload.
export async function loadTaskOwners() {
  const db = await readDb()
  const owners = {}
  for (const [taskId, value] of Object.entries(db.taskOwners || {})) {
    const subject = typeof value === 'string' ? value : value?.subject
    if (typeof subject === 'string' && subject.trim()) owners[String(taskId)] = subject.trim()
  }
  return owners
}

export function rememberTaskOwner(taskId, subject) {
  const id = String(taskId || '').trim()
  const owner = String(subject || '').trim()
  if (!id || !owner) return Promise.resolve()
  taskOwnerWrite = taskOwnerWrite.then(async () => {
    await withDbMutation(async (db) => {
      db.taskOwners[id] = { subject: owner, updatedAt: Math.floor(Date.now() / 1000) }
      const entries = Object.entries(db.taskOwners)
      if (entries.length > 10000) {
        entries.sort((a, b) => Number(a[1]?.updatedAt || 0) - Number(b[1]?.updatedAt || 0))
        for (const [staleId] of entries.slice(0, entries.length - 10000)) delete db.taskOwners[staleId]
      }
    })
  })
  return taskOwnerWrite
}

export function isManaged() {
  // An explicitly supplied placeholder must not silently select the local,
  // unauthenticated mode.  Keep the process managed and let the credential
  // validators fail closed with a 503 until real values are configured.
  return Boolean(String(process.env.SUB2API_SSO_SECRET || '').trim() || String(process.env.SUB2API_APP_CREDENTIAL || '').trim())
}

export function relayBaseUrl() {
  let raw = String(process.env.SUB2API_RELAY_BASE_URL || process.env.LINK || '').trim()
  if (raw && !raw.includes('://')) raw = `http://${raw}`
  raw = raw.replace(/\/$/, '')
  return raw.endsWith('/v1') ? raw : raw ? `${raw}/v1` : ''
}

export function satelliteHeaders(subject, { required = true } = {}) {
  const credential = String(process.env.SUB2API_APP_CREDENTIAL || '').trim()
  const user = String(subject || '').trim()
  if (required && (!hasConfiguredSecret(credential) || !user)) throw Object.assign(new Error('Sub2API satellite credentials are not configured'), { status: 503 })
  if (!hasConfiguredSecret(credential) || !user) return {}
  return {
    Authorization: `Bearer ${credential}`,
    'X-Sub2API-On-Behalf-Of': user,
    'X-Sub2API-Satellite': 'ai3d',
  }
}

export function sub2apiV1Base() {
  let base = String(process.env.SUB2API_RELAY_BASE_URL || process.env.LINK || '').trim().replace(/\/+$/, '')
  if (base && !/^https?:\/\//i.test(base)) base = `http://${base}`
  return base ? (base.endsWith('/v1') ? base : `${base}/v1`) : ''
}

export function sub2apiPurchaseUrl() {
  const raw = String(process.env.LINK || '').trim()
  if (!raw) return null
  try {
    const url = new URL(/^https?:\/\//i.test(raw) ? raw : `http://${raw}`)
    if (!['http:', 'https:'].includes(url.protocol) || url.username || url.password) return null
    url.pathname = '/purchase'
    url.search = ''
    url.hash = ''
    return url.toString().replace(/\/$/, '')
  } catch {
    return null
  }
}

export { encodePart }
